numpy/core/src/multiarray/ctors.c

```diff
@@ -985,14 +985,6 @@ PyArray_NewFromDescr_int(
     for (i = 0; i < nd; i++) {
         npy_intp dim = dims[i];
 
-        if (dim == 0) {
-            /*
-             * Compare to PyArray_OverflowMultiplyList that
-             * returns 0 in this case.
-             */
-            continue;
-        }
-
```

Comment:
To some extent this might be deliberate - it means that `arr[:,i].data` still points within allocated memory, rather than being an invalid pointer.

Another way to ensure this would be to just set all the strides to zero.

Perhaps it doesn't matter.

Reply:
nbytes is only used for new allocations, not for views.

Reply:
Right, but the point is that when I do `view = arr[:,i]` I end up with `view.data = arr.data + i * arr.strides[1]`. The construction here ensures that `view.data` lies within the chunk of memory allocated by `arr.data`.

Reply:
I don't see how this PR changes anything. The loop in question only is for `nbytes`, and `nbytes` is only used for new arrays, not for views. What am I missing?

Reply:
Let's imagine that arrays have a new `allocation_size` attribute, which is the number passed to `malloc`.
```
>>> a = np.zeros((2, 3, 0), np.uint8)
>>> a.allocation_size
6  # before this patch
1  # after this patch
>>> a.strides
(3, 1, 1)
>>> b = a[1:]
>>> diff = b.ctypes.data - a.ctypes.data
3  # same before and after this patch
>>>0 <= diff <= a.allocation_size
True  # before this patch
False  # after this patch
```

Unfortunately, if that last condition is false then we invoked undefined behavior at the point we computed `b.ctypes.data`

Reply:
`b = a[1:]` does nothing with `nbytes`. The code skips over the allocation stanza since `data != NULL`. I checked in a debugger.

Reply:
refactored the code to make the use of `nbytes` clearer

Reply:
Yeah, I think we may have actually said that this may be just as well when putting this in, hmmm (I do not remember). It seems a bit of a tiny issue: what is the use-case for huge but empty arrays, when the non-empty version would be too big to handle?
In fact, I think Eric is right, you could have to arrays which may appear to `may_share_memory`, unless we have a special case that empty arrays cannot share memory. Maybe it would be better to use `NULL` as the data pointer for empty arrays if we are going here, I hope that would not kill any users, but at least it would be obvious if there is incorrect use?

Reply:
@mattip, the issue is not what `a[1:]` does, but that as a result of your change the allocation within `a` is now dangerously small, and the bits of numpy that handle computing views invoke undefined behavior.

Reply:
Sorry, yeah, NULL would require way too many special cases.

Reply:
> I think we may have actually said that this may be just as well when putting this in

Note the offending code has been around for a long time, although github is serving me unicorns if I attempt to go back further: https://github.com/numpy/numpy/blob/91be9a9ace3e532a40a93f0c00d71fa34adce009%5E/numpy/core/src/arrayobject.c#L4618

Reply:
Hmm, true, so the actual change is just that there are more copies of these unused mallocs around? Also `_array_fill_strides` which used to be use, has the same behaviour since the dawn of time.

Reply:
The refactor fails, since we apparently need the error to be raised even for views.

Reply:
@eric-wieser could you expand on "the allocation within `a` is now dangerously small" ? As I understand it, either `b` is a view into `a`, in which case its `data` is NULL, or `b` has its own memory. The only case where `b.base` and `b.data` both are non-NULL is for `WRITEBACKIFCOPY`.

Reply:
Data is always non-null, but we normally make sure it points to valid memory even for empty arrays. With this change, that will not be true anymore, e.g.:
```
In [1]: arr = np.ones((2**59, 0), dtype=np.uint8)                                                                       
In [2]: arr2 = np.ones(100)                                                                                             
In [3]: diff = arr2.__array_interface__["data"][0] - arr.__array_interface__["data"][0]                                 

In [4]: arr[diff:].__array_interface__["data"][0]                                                                       
Out[4]: 24742928

In [5]: arr2.__array_interface__["data"][0]                                                                             
Out[5]: 24742928
```
`arr[diff:]` has nothing to do with `arr2`, but can be made to point to its memory. Now of course nobody should be accessing `arr[diff:]`'s memory. `np.may_share_memory` seems to survive this, although I wonder if that is due to some fastpath and not an actual general behaviour.

Reply:
Ahh, I see, the problem is when allocating `a` not when getting a view into it. The code in master will allocate "as if" the `0` values in the shape were replaced with `1`, so `np.zeros((0, 1000, 1000, 1000, 1000))` throws the error `MemoryError: Unable to allocate 0 bytes for an array with shape (0, 1000, 1000, 1000, 1000) and data type float64`. Now when taking a view into this, `b.ctypes.data` will point to somewhere inside the allocation. Is there a way to actually reference the memory allocated? Any `__getitem__` should fail. Using `ctypes.cast` is cheating, since I could just cast any integer to a pointer using that construct.

Reply:
> Is there a way to actually reference the memory allocated?

It's not relevant, because the C standard considers the following undefined behavior even though no dereference occurs:
```C
int x[4] = {1, 2, 3, 4];
int *end = x + 4;  // ok
int *past_end = x + 5;  // undefined behavior here
```

Reply:
`np.may_share_memory` has a [fast path](https://github.com/numpy/numpy/blob/v1.18.2/numpy/core/src/common/mem_overlap.c#L668) for `if dims[i] == 0`. Don't we consider places that allow access to memory beyond `strides[i] * dims[i]` to be a bug?

Reply:
Ah, I was not aware that the C-standard complains about it. If that is the case, there is probably no point in spending too much time on this? It will probably just work TM, but I honestly do not see the big issue with the current behaviour (or in other words: is there an actual use-case?)

Reply:
@seberg: "current" meaning master or this patch?

Reply:
Master behaviour

Reply:
The fact that `a.ctypes.data != b.ctypes.data` in your example above seems like another bug. If `a.size == 0` then I would expect any view of `a` to point to the same, if not NULL, `data`

Reply:
I'm not sure I'd consider it a bug, but option 2 from https://github.com/numpy/numpy/pull/15780#discussion_r395122076 would make the pointers equal.

Reply:
fixed so `a.ctypes.data == b.ctypes.data` if `any(a.shape ==0)` in 39e1233

Reply:
OK, I am not against this. One thing though: Some operation (such as reductions along the empty axis) want to combine the other axes (for iteration). The current check ensures that the total number of elements is limited by the size of `intp`. Removing this will remove that check, which means that such "reshapes" can fail. Maybe we can just ignore that, but certain operations will fail on such too-large arrays.